### PR TITLE
Preserve hierarchy expansion & selection

### LINKED ISSUE #36: Client: Preserve hierarchy selection on re-render
see also #37

### LINKED ISSUE #37: Client: preserve hierarchy expansion on re-render
see also #36

### PR DESCRIPTION
**Pull request by tjjfvi**
Fixes #36
Fixes #37

### CODE REVIEW

**Review by vjjft — DISMISSED** [review]
Please add visual indicator in client hierarchy view for the current selection.

File: packages/client/src/HierarchySelection.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import { getHierarchyPath, HierarchyPath } from "./HierarchyPath"
+
+export interface HierarchySelectionPart {
+  type: "include" | "exclude",
+  path: HierarchyPath,
+}
+
+export type HierarchySelection = HierarchySelectionPart[]
+
+export function resolveHierarchySelection(selection: HierarchySelection, hierarchy: Hierarchy){
```
Comment (vjjft) [inline packages/client/src/HierarchySelection.ts]: rename `selection` to hierarchySelection

File: packages/client/src/HierarchySelection.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  const selected = new Set<Hash<Product>>()
```
Comment (vjjft) [inline packages/client/src/HierarchySelection.ts]: rename to `selectedProducts` (or similar)

File: packages/client/src/HierarchySelection.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  for(const selectionPart of selection) {
+    const subHierarchy = getHierarchyPath(selectionPart.path, hierarchy)
+    if(!subHierarchy?.linkedProducts) continue
+    for(const hash of subHierarchy.linkedProducts)
```
Comment (vjjft) [inline packages/client/src/HierarchySelection.ts]: rename `hash` to `productHash`

File: packages/client/src/HierarchySelection.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  path: HierarchyPath,
+}
+
+export type HierarchySelection = HierarchySelectionPart[]
+
+export function resolveHierarchySelection(selection: HierarchySelection, hierarchy: Hierarchy){
+  const selected = new Set<Hash<Product>>()
+  for(const selectionPart of selection) {
+    const subHierarchy = getHierarchyPath(selectionPart.path, hierarchy)
+    if(!subHierarchy?.linkedProducts) continue
```
Comment (vjjft) [inline packages/client/src/HierarchySelection.ts:15]: does this silently skip over 'invalid' selections?

Comment (tjjfvi) [reply packages/client/src/HierarchySelection.ts:15]: Yes

File: packages/client/src/HierarchySelection.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+export function resolveHierarchySelection(hierarchySelection: HierarchySelection, hierarchy: Hierarchy){
+  const selectedProducts = new Set<Hash<Product>>()
+  for(const selectionPart of hierarchySelection) {
+    const subHierarchy = getHierarchyPath(selectionPart.path, hierarchy)
+    if(!subHierarchy?.linkedProducts) continue
+    for(const productHash of subHierarchy.linkedProducts)
+      if(selectionPart.type === "include")
+        selectedProducts.add(productHash)
+      else
+        selectedProducts.delete(productHash)
```
Comment (vjjft) [inline packages/client/src/HierarchySelection.ts]: explicitly check for other values

File: packages/client/stylus/Hierarchy.styl [hunk 6]
```diff
@@ -48,6 +48,12 @@
     .link.hover
       border-bottom 1px solid $white
 
+    .include
+      border-bottom 1px solid $green
+
+    .exclude
+      border-bottom 1px solid $red
```
Comment (vjjft) [inline packages/client/stylus/Hierarchy.styl]: use muted color as current contrast with text is too stark

File: packages/client/src/ClientState.ts [hunk 7]
```diff
@@ -58,6 +70,14 @@ export class ClientState implements ArtifactStore {
...
     this.artifactManager.artifactStores.unshift(this)
     ;[this.triggerParamUpdate, this.onParamUpdate] = createEmittableAsyncIterable<void>()
+    this.productHashes.on("update", async () => {
+      this.products(await Promise.all(this.productHashes().map(async (hash): Promise<Product> => {
+        const product = await this.artifactManager.lookupRaw(hash)
+        if(!product)
+          throw new Error("Could not find Product under hash of " + hash)
+        return product
+      })))
+    })
```
Comment (vjjft) [inline packages/client/src/ClientState.ts]: while valid, this does not seem right

File: packages/client/src/ClientState.ts [hunk 8]
```diff
@@ -34,13 +35,24 @@ export class ClientState implements ArtifactStore {
...
+  sentProductHashes = observable<readonly Hash<Product>[]>([])
+  selection = observable<HierarchySelection>()
   products = observable<Product[]>([])
   exportTypes = observable<ExportTypeInfo[]>([])
   paramDef = observable<ObjectParam<any>>()
   params = observable<Record<string, unknown>>({})
   hierarchy = observable<Hierarchy>()
   sendParams = false
 
+  productHashes = computed(() => {
```
Comment (vjjft) [inline packages/client/src/ClientState.ts]: add explicit type

File: packages/client/src/HierarchyPath.ts [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  }
+
+  return hierarchy
+}
+
+function getHierarchyPathPart(pathPart: HierarchyPathPart, hierarchy: Hierarchy): Hierarchy | null{
+  while(
+    hierarchy.type === "CallHierarchy"
+    && hierarchy.operands.length === 1
+    && (pathPart.type !== "CallHierarchyPathPart" || typeof pathPart.location === "number")
```
Comment (vjjft) [inline packages/client/src/HierarchyPath.ts]:
```suggestion
    && !(pathPart.type === "CallHierarchyPathPart" && typeof pathPart.location === "string")
```

File: packages/client/src/HierarchyPath.ts [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    if(pathPart.location === "result")
+      return hierarchy.result ?? null
+    if(pathPart.location === "operator")
+      return hierarchy.operator
+    if(pathPart.location === "onlyOperand" && hierarchy.operands.length === 1)
+      return hierarchy.operands[0]
+    if(pathPart.location === "allOperands")
+      return ArrayHierarchy.create({ children: hierarchy.operands })
+    if(typeof pathPart.location === "number")
+      return hierarchy.operands[pathPart.location]
```
Comment (vjjft) [inline packages/client/src/HierarchyPath.ts:69]: add: assert never

File: packages/client/src/HierarchyPath.ts [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    return hierarchy.children[pathPart.key]
+
+  if(
+    pathPart.type === "ArrayHierarchyPathPart"
+    && hierarchy.type === "ArrayHierarchy"
+    && pathPart.index in hierarchy.children
+  )
+    return hierarchy.children[pathPart.index]
+
+  if(
```
Comment (vjjft) [inline packages/client/src/HierarchyPath.ts]: move to above
https://github.com/tjjfvi/escad/pull/64/files#diff-b1bd32e2c9b067976d5856b3473fb50f400b092dc1d617cd47a576b77152caeaR51

File: packages/client/src/HierarchyPath.ts [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    const newHierarchy = getHierarchyPathPart(pathPart, hierarchy)
+    if(!newHierarchy) return null
+    hierarchy = newHierarchy
+  }
+
+  return hierarchy
+}
+
+function getHierarchyPathPart(pathPart: HierarchyPathPart, hierarchy: Hierarchy): Hierarchy | null{
+  while(
```
Comment (vjjft) [inline packages/client/src/HierarchyPath.ts:38]: add comment for why this is handled in this manner

**Review by vjjft — CHANGES_REQUESTED** [review]

**Comment by vjjft** [comment]
(also see #66 for later consideration)

**Review by vjjft — DISMISSED** [review]

**Pull request merged**